core/win32/syscall.c

```diff
@@ -2961,7 +2964,12 @@ pre_system_call(dcontext_t *dcontext)
     reg_t *param_base = pre_system_call_param_base(mc);
     dr_where_am_i_t old_whereami = dcontext->whereami;
     dcontext->whereami = DR_WHERE_SYSCALL_HANDLER;
-    IF_X64(ASSERT_TRUNCATE(sysnum, int, mc->xax));
+    // FIXME
+    // commenting this ASSERT since x64 client injected in
+    // WOW64 binaries will be using mc->rax as comparaison
+    // thus making the assert fail, while the syscall still
+    // being valid.
```

Comment:
A) Who is setting the upper bits?

B) You mean the Windows kernel ignores the upper bits?  So the assert is flawed?

Reply:
I meant that the assert always fails due to the size of mc->rax being compared with wrong sizes, no matter if the syscall is valid or not.

Reply:
I don't understand "the assert always fails due to the size of mc->rax being compared with wrong sizes"?

Reply:
Trying to print erroneous syscalls (syscalls that makes the assert trigger) on WoW64 + inject_x64, I can see that `mc->xax` value is `ffffffff00031054`. I assume that this comes from a conversion somehow representing the syscall with a signed number, which outputs a value that is wrong if you display it. Since `mc->xax` is only used as a sysnumber and type converted to `int`, this is not an issue, but this still makes `ASSERT_TRUNCATE` (more precisely `CHECK_TRUNCATE_TYPE_int`) fail.

Reply:
There would not be a widening conversion for an application register.  App register values have to be precisely preserved.  The machine context structure stores 64-bit values and prints them as such.  If the top bits there are 1's, either the app set them to 1's, or somehow there is a bug where DR corrupted the register value.  Is this for 32-bit app code?  If it is 64-bit, we can't put it past Windows making some change to have the top bits mean something -- though in the past 64-bit code wrote to `eax`, zeroing the top bits.